Add immutable dictionary and immutable hashset serializer

DrSmugleaf opened the pull request.

DrSmugleaf committed:
Add immutable dictionary and immutable hashset serializer

DrSmugleaf committed:
Reset unnecessary changes

DrSmugleaf committed:
Missing change

DrSmugleaf committed:
Rider bro

PJB3005 reviewed (changes requested):
Most of this code is copy-pasted from `DictionarySerializer` and `HashSetSerializer`. Please avoid that.

DrSmugleaf committed:
Deduplicate code

DrSmugleaf commented:
> Most of this code is copy-pasted from `DictionarySerializer` and `HashSetSerializer`. Please avoid that.

Fixed in [`2443d67` (#5)](https://github.com/space-wizards/netserializer/pull/5/commits/2443d67c86d9c640c208cc679c53ff531aec63ce), I misremembered and thought we shouldn't change existing code from the original netserializer repo

DrSmugleaf committed:
Fix compile

PJB3005 commented:
> I misremembered and thought we shouldn't change existing code from the original netserializer repo

Yeah it's fine. If we're doing anything major with NetSerializer it's rewriting it from scratch anyways (for multiple reasons).

PJB3005 commented on NetSerializer/Serializer.cs:
Avoid these formatting changes.

DrSmugleaf replied:
Fixed in [`0fa0d4d` (#5)](https://github.com/space-wizards/netserializer/pull/5/commits/0fa0d4d2ef64712e7a69ee1bb75b7897401743b5)

PJB3005 reviewed (changes requested):
Aren't you missing the write implementations, and the read implementations have the wrong name so the existing code won't locate them?

This doesn't work for me.

```cs
        [Test]
        public void TestImmutableDictionary()
        {
            var x = new Dictionary<string, int>
            {
                { "A", 1 },
                { "B", 2 }
            }.ToImmutableDictionary();

            var serializer = new Serializer([typeof(ImmutableDictionary<string, int>)], new Settings());
            var stream = new MemoryStream();
            serializer.SerializeDirect(stream, x);
            stream.Position = 0;
            serializer.DeserializeDirect(stream, out ImmutableDictionary<string, int> read);

            Assert.That(read, NUnit.Framework.Is.EquivalentTo(x));
        }
```

```
System.NullReferenceException : Object reference not set to an instance of an object.
   at NetSerializer.DictionarySerializer.GetStaticWriter(Type type) in E:\ss14\space-station-14\RobustToolbox\NetSerializer\NetSerializer\TypeSerializers\DictionarySerializer.cs:line 53
   at NetSerializer.Serializer.GenerateWriterStub(Type type) in E:\ss14\space-station-14\RobustToolbox\NetSerializer\NetSerializer\Serializer.cs:line 496
   at NetSerializer.Serializer.GenerateWriters(Type rootType) in E:\ss14\space-station-14\RobustToolbox\NetSerializer\NetSerializer\Serializer.cs:line 550
   at NetSerializer.Serializer.GenerateDirectWriterDelegate(Type type) in E:\ss14\space-station-14\RobustToolbox\NetSerializer\NetSerializer\Serializer.cs:line 580
   at NetSerializer.Serializer.SerializeDirect[T](Stream stream, T value) in E:\ss14\space-station-14\RobustToolbox\NetSerializer\NetSerializer\Serializer.cs:line 352
```

DrSmugleaf committed:
rider pls

DrSmugleaf committed:
Fix method names deduplicate code

DrSmugleaf commented:
> Aren't you missing the write implementations, and the read implementations have the wrong name so the existing code won't locate them?

Fixed in [`76e2b25` (#5)](https://github.com/space-wizards/netserializer/pull/5/commits/76e2b25f21c4804bda75404e221ac0a16ed42a88), added tests in https://github.com/space-wizards/RobustToolbox/pull/6109

DrSmugleaf committed:
Fix missing optional argument
